<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Cola Sequence Aligner : Modular and Configurable Optimal Sequence Aligner">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Cola Sequence Aligner</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/nedaz/cola">View on GitHub</a>

          <h1 id="project_title">Cola Sequence Aligner</h1>
          <h2 id="project_tagline">Modular and Configurable Optimal Sequence Aligner</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/nedaz/cola/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/nedaz/cola/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>Cola consists of an efficient implementation of a collection of sequence
alignment algorithms, extending the Smith-Waterman and Smith-Waterman-Gap-Affine
methods by the ability to apply a scoring function that is an arbitrary
function of the size of consecutive nucleotide matches.</p>

<hr>

<hr>

<p>If you want to integrate Cola with your code, that is simple. Look into src/runCola as
an example of how the Cola aligner object can be used.</p>

<hr>

<p>Once you unzip the code-base, run make in the parent directory. Then you will have the 
executables created and will be able to follow the instructions below:</p>

<hr>

<p>The main access point for a user to run cola is through runCola which accepts the following arguments:</p>

<ul>
<li>-q  : Query sequence in FASTA format</li>
<li>-t  : Target sequence in FASTA format</li>
<li>-a  : Aligner type - Choose a number from 1 to 4 for the following modes - 1 : NSGA , 2 : NS , 3 : SWGA, 4 : SW </li>
<li>-o  : Aligner gap open penalty</li>
<li>-m  : Aligner mismatch penalty</li>
<li>-e  : Aligner gap extension penalty</li>
<li>-all: align all to all sequences (default is false)</li>
<li>-p : Maximum acceptable P-value (def=1)</li>
<li>-i : Minium acceptable identity (def=0)</li>
<li>-b : The bandwidth for banded mode, default is for unbanded (def=-1)</li>
</ul>

<hr>

<p>N.B. Only the first 3 arguments are compulsory and the rest are optional. 
If the aligner penalty parameters are not provided by user, defaults will be used. </p>

<hr>

<p>See following examples for more detail:</p>

<hr>

<p>To run in one of 4 modes using the sample data provided in the sample directory:</p>

<hr>

<ol>
<li><p>Nonlinear Scoring Gap-Affine (NSGA): 
./runCola -t sample/homo.X.part.fa -q sample/canis.X.part.fa -a 1 </p></li>
<li><p>Nonlinear Scoring (NS): 
./runCola -t sample/homo.X.part.fa -q sample/canis.X.part.fa -a 2 </p></li>
<li><p>Smith-Waterman Gap-Affine (SWGA): 
./runCola -t sample/homo.X.part.fa -q sample/canis.X.part.fa -a 3 </p></li>
<li><p>Smith-Waterman (SW): 
./runCola -t sample/homo.X.part.fa -q sample/canis.X.part.fa -a 4 </p></li>
<li><p>See Cola manuscript for more detail on the underlying algorithm and methodology.</p></li>
</ol>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Cola Sequence Aligner maintained by <a href="https://github.com/nedaz">nedaz</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
